src/Apps/W1/External Storage - Document Attachments/app/src/AutomaticSync/DAExternalStorageSync.Report.al

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+
+namespace Microsoft.ExternalStorage.DocumentAttachments;
+
+using Microsoft.Foundation.Attachment;
+
+/// <summary>
+/// Report for synchronizing document attachments between internal and external storage.
+/// Supports bulk upload, download, and cleanup operations.
+/// </summary>
+report 8752 "DA External Storage Sync"
+{
+    Caption = 'External Storage Synchronization';
+    ProcessingOnly = true;
+    UseRequestPage = true;
+    Extensible = false;
+    UsageCategory = None;
+    Permissions = tabledata "DA External Storage Setup" = r,
+                  tabledata "Document Attachment" = r;
+
+    dataset
+    {
+        dataitem(DocumentAttachment; "Document Attachment")
+        {
+            trigger OnPreDataItem()
+            begin
+                SetFilters();
+                TotalCount := Count();
+
+                if TotalCount = 0 then begin
+                    if GuiAllowed() then
+                        Message(NoRecordsMsg);
+                    CurrReport.Break();
+                end;
+
+                if MaxRecordsToProcess > 0 then
+                    if TotalCount > MaxRecordsToProcess then
+                        TotalCount := MaxRecordsToProcess;
+
+                ProcessedCount := 0;
+                FailedCount := 0;
+                DeleteCount := 0;
+                DeleteFailedCount := 0;
+
+                if GuiAllowed() then
+                    Dialog.Open(ProcessingMsg, TotalCount);
+            end;
+
+            trigger OnAfterGetRecord()
+            var
+                DateFormulaLbl: Label '<+%1>', Locked = true;
+            begin
+                ProcessedCount += 1;
+
+                if GuiAllowed() then
+                    Dialog.Update(1, ProcessedCount);
+
+                case SyncDirection of
+                    SyncDirection::"To External Storage":
+                        if not ExternalStorageProcessor.UploadToExternalStorage(DocumentAttachment) then
+                            FailedCount += 1;
+                    SyncDirection::"From External Storage":
+
+                        if not ExternalStorageProcessor.DownloadFromExternalStorage(DocumentAttachment) then
+                            FailedCount += 1;
+                end;
+                if DeleteExpiredFiles then
+                    if CalcDate(StrSubstNo(DateFormulaLbl, GetDateFormulaFromExternalStorageSetup()), DocumentAttachment."External Upload Date".Date()) >= Today() then
+                        if ExternalStorageProcessor.DeleteFromInternalStorage(DocumentAttachment) then
+                            DeleteCount += 1
+                        else
+                            DeleteFailedCount += 1;
```

Comment:
Should we maybe have an commit after each record?
Since we communicate to an external system it's very likely that we receive an error in a large batch.
The files will have been moved to the exteranl system, therefore we shouldn't lost the information about the changes on the document attachment.